mm2src/coins/eth/web3_transport/websocket_transport.rs

```diff
@@ -0,0 +1,343 @@
+                                             let response_map = unsafe { &mut *self.responses.0 };
+                                             let _ = response_map.insert(request_id, res_bytes);
+
+                                             notifier.send(()).expect("receiver channel must be alive");
+                                         }
+
+                                     }
+                                 }
+                             }
+                         },
+                         Some(Ok(tokio_tungstenite_wasm::Message::Binary(_))) => continue,
+                         Some(Ok(tokio_tungstenite_wasm::Message::Close(_))) => break,
+                         Some(Err(e)) => {
+                            log::error!("{e}");
+                            return;
+                         },
+                         None => continue,
+                    }
+                }
+            }
+        }
+    }
+
+    pub(crate) async fn stop_connection_loop(&self) {
+        let mut tx = self.controller_channel.tx.lock().await;
+        tx.send(ControllerMessage::Close)
+            .await
+            .expect("receiver channel must be alive");
+
+        let response_map = unsafe { &mut *self.responses.0 };
+        response_map.clear();
+    }
+
+    pub(crate) fn maybe_spawn_connection_loop(&self, coin: EthCoin) {
+        // if we can acquire the lock here, it means connection loop is not alive
+        if self.connection_guard.try_lock().is_some() {
+            let fut = self.clone().start_connection_loop();
+            let settings = AbortSettings::info_on_abort(format!("connection loop stopped for {:?}", self.node.uri));
+            coin.spawner().spawn_with_settings(fut, settings);
+        }
+    }
+}
+
+async fn send_request(
+    transport: WebsocketTransport,
+    request: Call,
+    request_id: RequestId,
+    event_handlers: Vec<RpcTransportEventHandlerShared>,
+) -> Result<serde_json::Value, Error> {
+    let mut serialized_request = to_string(&request);
+
+    if transport.node.gui_auth {
+        match handle_gui_auth_payload(&transport.gui_auth_validation_generator, &request) {
+            Ok(r) => serialized_request = r,
+            Err(e) => {
+                return Err(Error::Transport(TransportError::Message(format!(
+                    "Couldn't generate signed message payload for {:?}. Error: {e}",
+                    request
+                ))));
+            },
+        };
+    }
+
+    let mut tx = transport.controller_channel.tx.lock().await;
+
+    let (notification_sender, notification_receiver) = futures::channel::oneshot::channel::<()>();
+
+    event_handlers.on_outgoing_request(serialized_request.as_bytes());
+
+    tx.send(ControllerMessage::Request(WsRequest {
+        request_id,
+        serialized_request,
+        response_notifier: notification_sender,
+    }))
+    .await
+    .expect("receiver channel must be alive");
```

Comment:
I think its worth propagating the error here instead of panic as `send_request` function returns result.